include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -0,0 +1,1386 @@
+#ifdef PREFETCH_REG_COUNT_1
+			if (regcnt.y==1u)
+				writeReg(reg, val.x);
+			else
+#endif
+#ifdef PREFETCH_REG_COUNT_2
+			if (regcnt.y==2u)
+				writeReg(reg, val.xy);
+			else
+#endif
+#ifdef PREFETCH_REG_COUNT_3
+			if (regcnt.y==3u)
+				writeReg(reg, val);
+			else
+#endif
+			{} //else "empty braces"
+		}
+#endif
+#ifdef PREFETCH_TEX_2
+		if ((fetchFlags & 0x04u) == 0x04u)
+		{
+			uvec3 texid = uvec3(bsdf_data.data[1].zw, bsdf_data.data[2].x);
+			vec3 val = fetchTex(texid,UV,dUV);
+			uint reg = REG_PREFETCH2(regs);
+#ifdef PREFETCH_REG_COUNT_1
+			if (regcnt.z==1u)
+				writeReg(reg, val.x);
+			else
+#endif
+#ifdef PREFETCH_REG_COUNT_2
+			if (regcnt.z==2u)
+				writeReg(reg, val.xy);
+			else
+#endif
+#ifdef PREFETCH_REG_COUNT_3
+			if (regcnt.z==3u)
+				writeReg(reg, val);
+			else
+#endif
+			{} //else "empty braces"
+		}
+#endif
+#ifdef PREFETCH_TEX_3
+		if ((fetchFlags & 0x08u) == 0x08u)
+		{
+			vec3 val = fetchTex(bsdf_data.data[2].yzw, UV, dUV);
+			uint reg = REG_PREFETCH3(regs);
+#ifdef PREFETCH_REG_COUNT_1
+			if (regcnt.w==1u)
+				writeReg(reg, val.x);
+			else
+#endif
+#ifdef PREFETCH_REG_COUNT_2
+			if (regcnt.w==2u)
+				writeReg(reg, val.xy);
+			else
+#endif
+#ifdef PREFETCH_REG_COUNT_3
+			if (regcnt.w==3u)
+				writeReg(reg, val);
+			else
+#endif
+			{} //else "empty braces"
+		}
+#endif
+	}
+}
+
+void runNormalPrecompStream(in instr_stream_t stream, in mat2 dUV)
+{
+	instr_execute_SET_GEOM_NORMAL();
+	for (uint i = 0u; i < stream.count; ++i)
+	{
+		instr_t instr = irr_glsl_MC_fetchInstr(stream.offset+i);
+
+		bsdf_data_t bsdf_data = fetchBSDFDataForInstr(instr);
+
+		uint srcreg = bsdf_data.data[0].z;
+		uint dstreg = REG_DST(instr_decodeRegisters(instr));
+
+		vec2 dh = readReg2(srcreg);
+		
+		writeReg(dstreg,
+			irr_glsl_perturbNormal_derivativeMap(currInteraction.isotropic.N, dh, currInteraction.isotropic.V.dPosdScreen, dUV)
+		);
+	}
+}
+
+void flipCurrInteraction()
+{
+	currInteraction.isotropic.N = -currInteraction.isotropic.N;
+	currInteraction.isotropic.NdotV = -currInteraction.isotropic.NdotV;
+	currInteraction.T = -currInteraction.T;
+	currInteraction.B = -currInteraction.B;
+	currInteraction.TdotV = -currInteraction.TdotV;
+	currInteraction.BdotV = -currInteraction.BdotV;
+}
+//call before executing an instruction/evaluating bsdf
+void handleTwosided_interactionOnly(inout bool ts_flag, in instr_t instr)
+{
+#ifndef NO_TWOSIDED
+	if (!ts_flag && instr_getTwosided(instr))
+	{
+		ts_flag = true;
+		if (currInteraction.isotropic.NdotV<0.0)
+		{
+			flipCurrInteraction();
+		}
+	}
+#ifdef OP_BUMPMAP
+	ts_flag = instr_getOpcode(instr)==OP_BUMPMAP ? false:ts_flag;
+#endif
+#endif	
+}
+//call before executing an instruction/evaluating bsdf
+void handleTwosided(inout bool ts_flag, in instr_t instr, inout irr_glsl_LightSample s, inout irr_glsl_AnisotropicMicrofacetCache uf)
+{
+#ifndef NO_TWOSIDED
+	if (!ts_flag && instr_getTwosided(instr))
+	{
+		ts_flag = true;
+		if (currInteraction.isotropic.NdotV<0.0)
```

Comment:
you could roll this into one if-statement

Reply:
you also don't need the `ts_flag` to make sure you don't flip twice, because `currInteraction.isotropic.NdotV<0.0` ensures you won't

P.S. This also means you dont need any special handling for the bumpmap case

Reply:
however because "two sided" was only supposed to be applied to some BSDF nodes, you need to keep track of whether the directions were flipped, so I propose this

```glsl
void handleTwosided(in bool originallyWasBackface, in instr_t instr, inout irr_glsl_LightSample s, inout irr_glsl_AnisotropicMicrofacetCache uf)
{
   const bool twosided = instr_getTwosided(instr);
   const bool currentFrontface = currInteraction.isotropic.NdotV>0.0;
   if (twosided!=currentFrontface && originallyWasBackface)
   {
      do the flip
   }
}
```

100% guaranteed to work.

Reply:
ignoring because twosided-ness is going to be reworked anyway